internal/providers/exoscale/exoscale.go

```diff
@@ -37,10 +39,16 @@ var (
 
 // FetchConfig fetch Exoscale ign user-data config
 func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
-	data, err := f.FetchToBuffer(userdataURL, resource.FetchOptions{})
-	if err != nil && err != resource.ErrNotFound {
+	data, err := f.FetchToBuffer(userdataURL, resource.FetchOptions{
+		Compression: "gzip",
+	})
+	if err != nil && err != io.EOF && err != resource.ErrNotFound {
```

Comment:
This `io.EOF` check is probably not what we want but I wanted to start the conversation.

Basically the problem that we are having here is that `http://169.254.169.254/1.0/user-data` returns an empty file (0 bytes) and the [gzip.NewReader()](https://github.com/coreos/ignition/blob/919102e70689dfa2522861aeb131cae44b990352/internal/resource/url.go#L489) inside of [decompressCopyHashAndVerify()](https://github.com/coreos/ignition/blob/919102e70689dfa2522861aeb131cae44b990352/internal/resource/url.go#L500-L503) barfs on that.

I don't know the best way to check if the reader is zero length before calling gzip.NewReader(r) on it. Should we check first or is detecting the returned "EOF" from it good enough to assume the file/stream was empty?

Reply:
IOW `gzip.NewReader()` returns "EOF" in this particular case and I don't know how to properly handle that.

Reply:
EOF could mean a mid-stream EOF, which would be a legitimate error, so this test is indeed too coarse.

To check the underlying stream for EOF before wrapping it, we'd need to read ahead a byte, which we can do by wrapping the stream in a `bufio.Reader` and doing a `Peek`.  But that requires changes to `url.go` library code, and probably a `FetchOptions` flag, since in general a zero-byte gzip stream should be considered corrupt.

Instead, I think the cleanest approach is to fetch the data in compressed form, and then in the exoscale code, check whether it's empty and explicitly decompress.  This is less memory-efficient because we wouldn't be doing streaming decompression, but we should have RAM to spare.  If we needed `decompressCopyHashAndVerify` to check the config hash, this would be a lot less clean, but there's no config hash for a provider's root config.

Reply:
new update that does this now.